src/renderer/viz/expressions/RampGeneric.js

```diff
@@ -0,0 +1,168 @@
+
+        if (this.palette.type === paletteTypes.IMAGE_LIST) {
+            checkType('ramp', 'input', 0, inputTypes.CATEGORY, this.input);
+            checkInstance('ramp', 'palette', 1, ListImage, this.palette);
+        }
+
+        this.others._bindMetadata(metadata);
+        this.childrenNames.push('others');
+
+        this._properties = metadata.properties;
+        this._texCategories = null;
+        this._GLtexCategories = null;
+        this._metadata = metadata;
+    }
+
+    eval (feature) {
+        const input = this.input.eval(feature);
+        return this._calcEval(input, feature);
+    }
+
+    _calcEval (input, feature) {
+        const maxPaletteSize = this.input.numCategoriesWithoutOthers < this.palette.length
+            ? this.palette.length
+            : this.input.numCategoriesWithoutOthers;
+
+        const { palette, others } = this._getPalette();
+        const paletteValues = this.palette.isA(Palette)
+            ? palette.map((color) => color.eval(feature))
+            : this.palette.eval(feature).slice(0, maxPaletteSize);
+
+        if (input === OTHERS_INDEX) {
+            return others.eval(feature);
+        }
+
+        const maxValues = paletteValues.length - 1;
+        const min = Math.floor(input * maxValues);
+        const max = Math.ceil(input * maxValues);
+        const m = fract(input * maxValues);
+
+        return mix(paletteValues[min], paletteValues[max], m);
+    }
+
+    getLegend (options) {
+        const config = Object.assign({}, DEFAULT_OPTIONS, options);
+        const type = this.input.type;
+        const legendData = this.input.getLegendData(config);
+        const data = legendData.data.map(({key, value}) => {
+            value = this._calcEval(value, undefined);
+            return { key, value };
+        });
+
+        return { type, ...legendData, data };
+    }
+
+    _applyToShaderSource (getGLSLforProperty) {
+        const input = this.input._applyToShaderSource(getGLSLforProperty);
+        const { palette, others } = this._getPalette();
+        const GLSLPalette = palette.map(color => color._applyToShaderSource(getGLSLforProperty));
+        const GLSLOthers = others._applyToShaderSource(getGLSLforProperty);
+        const GLSLBlend = this.palette.type === 'number-list'
+            ? _getInlineGLSLBlend(GLSLPalette)
+            : _getInlineColorGLSLBlend(GLSLPalette);
+
+        const rampColorType = this.palette.type === 'number-list' ? 'float' : 'vec4';
+        const inline = `ramp_color${this._uid}(${input.inline})`;
+        const prefaceGLSL = `
+            ${CIELabGLSL}
+            ${GLSLPalette.map(elem => elem.preface).join('\n')}
+            ${GLSLOthers.preface}
+
+            ${rampColorType} ramp_color${this._uid}(float x){
+                return x==${OTHERS_GLSL_VALUE}
+                    ? ${GLSLOthers.inline}
+                    : ${GLSLBlend};
+            }`;
+
+        const preface = this._prefaceCode(input.preface + prefaceGLSL);
```

Comment:
to be consistent `input.preface` should be interpolated like `CIELabGLSL` is before.